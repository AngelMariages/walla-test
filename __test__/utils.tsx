import { render } from '@testing-library/react';
import { FavoritesContextProvider } from 'context/FavoritesContext';
import { FiltersContextProvider } from 'context/FiltersContext';

export const renderWithContext = (ui: React.ReactElement) => {
    return render(
        <FiltersContextProvider>
            <FavoritesContextProvider>{ui}</FavoritesContextProvider>
        </FiltersContextProvider>
    );
};
