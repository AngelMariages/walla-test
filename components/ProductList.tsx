import { fetchItems } from '../pages/api/items';
import Pagination from './Pagination';
import Product from './Product';

type Props = {
    search?: string;
};

function asyncComponent<T, R>(fn: (arg: T) => Promise<R>): (arg: T) => R {
    return fn as (arg: T) => R;
}

export default asyncComponent(async function ProductList({ search }: Props) {
    const { data } = await fetchItems(0, search);

    return (
        <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-4 bg-gray-400">
            {data.items.map((item, index) => (
                <Product key={index} item={item} />
            ))}
            {data.hasMore && <Pagination initialPage={1} search={search} />}
        </div>
    );
});
