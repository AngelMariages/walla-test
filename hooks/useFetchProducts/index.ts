import { useEffect, useState } from 'react';
import { Sort } from 'constants/sortBy';
import { fetchItemsOnClient } from 'lib/fetchItemsOnClient';
import { ApiResult } from 'pages/api/items';
import { Item } from 'constants/item';

const useFetchProducts = (
    page: number,
    sort: Sort,
    search?: string,
    initialData?: ApiResult,
) => {
    const [isLoading, setIsLoading] = useState(false);
    const [hasMore, setHasMore] = useState(false);
    const [products, setProducts] = useState<Item[]>(
        initialData?.data.items || []
    );

    useEffect(() => {
        (async () => {
            if (isLoading) return;

            setIsLoading(true);

            const { data } = await fetchItemsOnClient(page, search, sort);

            setProducts((prev) =>
                page === 1 ? [...data.items] : [...prev, ...data.items]
            );
            setHasMore(data.hasMore);

            setIsLoading(false);
        })();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [search, page, sort]);

    return {
        isLoading,
        products,
        hasMore,
    };
};

export default useFetchProducts;
