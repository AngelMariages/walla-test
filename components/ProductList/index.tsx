import { useCallback, useContext } from 'react';
import { FavoritesContext } from 'context/FavoritesContext';
import { FiltersContext } from 'context/FiltersContext';
import useFetchProducts from 'hooks/useFetchProducts';
import useInfinteScroll from 'hooks/useInfiniteScroll';
import { ApiResult } from 'pages/api/items';
import LoadingIcon from 'components/LoadingIcon';
import Product from 'components/Product';

type Props = {
    initialData?: ApiResult;
};

const ProductList: React.FC<Props> = ({ initialData }) => {
    const { search, page, sort, incrementPage } = useContext(FiltersContext);
    const { isFavorite } = useContext(FavoritesContext);

    const { isLoading, products, hasMore } = useFetchProducts(
        initialData,
        page,
        sort,
        search
    );

    const { itemRef } = useInfinteScroll(
        useCallback(() => {
            if (!isLoading) {
                incrementPage();
            }
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [isLoading])
    );

    return (
        <div className="w-full grid sm:grid-cols-1 md:grid-cols-2 lg:grid-cols-3 2xl:grid-cols-4 gap-4 bg-gray-300">
            {products.length === 0 ? (
                <div className="text-center text-2xl font-bold">
                    No items found
                </div>
            ) : null}
            {products.map((item, index) => (
                <Product key={index} item={item} selected={isFavorite(item)} />
            ))}
            {hasMore || isLoading ? (
                <div ref={itemRef}>
                    <LoadingIcon />
                </div>
            ) : null}
        </div>
    );
};

export default ProductList;
