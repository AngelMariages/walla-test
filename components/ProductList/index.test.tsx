import useFetchProducts from 'hooks/useFetchProducts';
import { PropsWithChildren } from 'react';
import { renderWithContext } from '__test__/utils';
import ProductList from '.';

jest.mock('hooks/useFetchProducts', () => {
    return jest.fn(() => ({
        isLoading: false,
        products: [],
        hasMore: false,
    }));
});

jest.mock('hooks/useInfiniteScroll', () => {
    return jest.fn(() => ({
        itemRef: null,
    }));
});

const mockProductComponent = jest.fn();

jest.mock('components/Product', () => {
    return function ProductComponent(props: PropsWithChildren) {
        mockProductComponent(props);
        return <div {...props}>{props.children}</div>;
    };
});

const mockLoadingIconComponent = jest.fn();

jest.mock('components/LoadingIcon', () => {
    return function LoadingIconComponent(props: PropsWithChildren) {
        mockLoadingIconComponent(props);
        return <div {...props}>{props.children}</div>;
    };
});

describe('ProductList', () => {
    it('should render correctly empty list', () => {
        const { container } = renderWithContext(<ProductList />);
        expect(container).toMatchSnapshot();
        expect(useFetchProducts).toHaveBeenCalledTimes(1);
        expect(useFetchProducts).toHaveBeenCalledWith(
            undefined,
            1,
            { sortBy: 'title', sortOrder: 'asc' },
            ''
        );
    });

    it('should render correctly with products', () => {
        const item = {
            title: 'title',
            price: 300,
            description: 'description',
            email: 'test@mail.com',
            image: 'image.jpg',
        };
        const item2 = {
            title: 'title2',
            price: 300,
            description: 'description',
            email: 'test@mail.com',
            image: 'image.jpg',
        };

        (useFetchProducts as jest.Mock).mockImplementation(() => {
            return {
                isLoading: false,
                products: [item, item2],
                hasMore: false,
            };
        });

        const { container } = renderWithContext(<ProductList />);
        expect(container).toMatchSnapshot();

        expect(mockProductComponent).toHaveBeenCalledTimes(2);
        expect(mockProductComponent).toHaveBeenCalledWith({
            item: item,
        });
        expect(mockProductComponent).toHaveBeenCalledWith({
            item: item2,
        });
    });

    it('should render loading icon if hasMore or isLoading', () => {
        (useFetchProducts as jest.Mock).mockImplementation(() => {
            return {
                isLoading: true,
                products: [],
                hasMore: false,
            };
        });

        const { container } = renderWithContext(<ProductList />);
        expect(container).toMatchSnapshot();

        expect(mockLoadingIconComponent).toHaveBeenCalledTimes(1);

        mockLoadingIconComponent.mockClear();

        (useFetchProducts as jest.Mock).mockImplementation(() => {
            return {
                isLoading: false,
                products: [],
                hasMore: true,
            };
        });

        const { container: container2 } = renderWithContext(<ProductList />);
        expect(container2).toMatchSnapshot();

        expect(mockLoadingIconComponent).toHaveBeenCalledTimes(1);
    });
});
