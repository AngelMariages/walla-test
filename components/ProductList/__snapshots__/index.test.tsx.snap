// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList should render correctly empty list 1`] = `
<div>
  <div
    class="w-full grid sm:grid-cols-1 md:grid-cols-2 lg:grid-cols-3 2xl:grid-cols-4 gap-4 bg-gray-300"
    data-testid="products-list"
  >
    <div
      class="text-center text-2xl font-bold"
    >
      No items found
    </div>
  </div>
</div>
`;

exports[`ProductList should render correctly with products 1`] = `
<div>
  <div
    class="w-full grid sm:grid-cols-1 md:grid-cols-2 lg:grid-cols-3 2xl:grid-cols-4 gap-4 bg-gray-300"
    data-testid="products-list"
  >
    <div
      data-testid="product-0"
      item="[object Object]"
    />
    <div
      data-testid="product-1"
      item="[object Object]"
    />
  </div>
</div>
`;

exports[`ProductList should render loading icon if hasMore or isLoading 1`] = `
<div>
  <div
    class="w-full grid sm:grid-cols-1 md:grid-cols-2 lg:grid-cols-3 2xl:grid-cols-4 gap-4 bg-gray-300"
    data-testid="products-list"
  >
    <div
      class="text-center text-2xl font-bold"
    >
      No items found
    </div>
    <div>
      <div />
    </div>
  </div>
</div>
`;

exports[`ProductList should render loading icon if hasMore or isLoading 2`] = `
<div>
  <div
    class="w-full grid sm:grid-cols-1 md:grid-cols-2 lg:grid-cols-3 2xl:grid-cols-4 gap-4 bg-gray-300"
    data-testid="products-list"
  >
    <div
      class="text-center text-2xl font-bold"
    >
      No items found
    </div>
    <div>
      <div />
    </div>
  </div>
</div>
`;
